### PR TITLE
E2E test framework

### PR DESCRIPTION
**Pull request by raghavendra-talur**
This PR adds
* e2e directory
* e2e config directory and a config file

We use viper to do the loading and validation of the config.

We have a new go.mod for the e2e dir and therefore we need go.work in
the top dir. This ensures that we keep the testing dependencies separate
from the Ramen dependencies. GO WORK is supported and recommended way of
doing this from Go version 1.18.

The sanity test suite has just one test in it and it checks if we get use the kubeconfigs to do a get namespace or not.

In the coming PRs, we can add new suites.

### CODE REVIEW

File: e2e/main_test.go [hunk 1]
```diff
@@ -72,6 +73,8 @@ func TestMain(t *testing.T) {
 
 	defer testContext.Cleanup()
 
+	suite.Run(t, &SanitySuite{testContext: testContext})
```
Comment (nirs) [inline e2e/main_test.go:76]: Don't use sanity, we just cleanup this term from ramen recently (#693).

Maybe call it BasicSuite? we can have ExtendedSuite later of maybe several suites 
focusing on specific configurations.

**Review by nirs — COMMENTED** [review]
I did not look at the code except one little comment, because I'm missing the big picture.

Can we start with a document (docs/e2e.md?) documenting how the framework work?

- how to run all tests
- how to select only some tests
- how to test with different configurations? (regional dr, metro dr)
  - can we use the same tests with both environments?
  - will we have tests that work only with some setup?
- how to run tests with existing env (not created by drenv)
- how to test with openshift (assuming the tests can run with openshift + upstream ramen)

I think we discussed this in the past,but we really need to make it clear what are we trying to build.

File: go.work [hunk 2]
```diff
@@ -0,0 +1,6 @@
+go 1.18
```
Comment (ShyamsundarR) [inline go.work:1]: Do we really need sub-modules? Can't we avoid the additional go workspace and have a single module with the e2e package in it?

**Comment by ShyamsundarR** [comment]
I expect to run this with drenv as follows:
- Create a drenv based on an environment file
- Pass the environment file as input to the e2e executable/main
- e2e reads the environment and decides what it supports (async/sync/volrep/volsync/...) and invokes the appropriate sub-tests on the environment

From a CI/D perspective;
- a jenkins (or similar) job would invoke in parallel(?) e2e tests against disjoint dr environments

For a non-dr environment, a kickstart file like the drenv environment file can potentially govern what is tested and what not? This maybe apart from [suites](https://github.com/stretchr/testify#suite-package) that we define within the tests for  narrower testing scope when needed?

File: e2e/main_test.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+type Config struct {
+	Clusters map[string]struct {
+		KubeconfigPath string `mapstructure:"kubeconfigpath" required:"true"`
+	} `mapstructure:"clusters" required:"true"`
+}
+type TestContext struct {
+	hubClient kubernetes.Interface
+	c1Client  kubernetes.Interface
+	c2Client  kubernetes.Interface
+}
```
Comment (nirs) [inline e2e/main_test.go:34]: Can we move TestContext to e2e/context package?

File: Makefile [hunk 4]
```diff
@@ -172,6 +172,9 @@ test-util-pvc: generate manifests envtest
 test-drenv:
 	$(MAKE) -C test
 
+test-e2e:
+	go test ./e2e
```
Comment (nirs) [inline Makefile:176]: When running locally you want to test one or more tests against one or more environments. How are we passing the environment?

Maybe this is not needed from the makefile, maybe it should run all the tests, and for running specific test you will use go test directly. However this must be documented so everyone can use it.

File: e2e/regional_DR_test.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package e2e_test
```
Comment (nirs) [inline e2e/regional_DR_test.go:1]: I would just call this example_test.go for now. We don't want to have all regional dr tests in one module, and most tests should be compatible with all environments, so we don't want to focus on the environment type in the test names.

File: e2e/utils_test.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	"k8s.io/client-go/kubernetes"
+)
+
+func runOnAllClusters(testContext *TestContext, f func(kubernetes.Interface) error) error {
+	var err error
+
+	err = f(testContext.HubClient())
+	if err != nil {
+		fmt.Printf("Failed to run function on hub cluster: %v\n", err)
+	}
```
Comment (nirs) [inline e2e/utils_test.go:15]: We want to support environment without a hub - drenv and ramenctl supports this now.

File: e2e/utils_test.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+	err = f(testContext.C2Client())
+	if err != nil {
+		fmt.Printf("Failed to run function on cluster 2: %v\n", err)
+	}
+
+	return err
+}
+
+func runOnAllOCPClusters(testContext *TestContext, f func(string) error) error {
```
Comment (nirs) [inline e2e/utils_test.go:30]: Not sure why this is so different from runOnAllClusters, and why the names are not consistent.

File: docs/e2e-test.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
+# Ramen End-to-End (E2E) Test Framework
+
+This document will provide an introduction to the framework, describe its components and structure, and explain how to run the tests.
+
+## Overview
+
+The E2E test framework is designed to validate the functionality of the Ramen project from a user's perspective. This means that it will test the entire system. The framework is built using Go and follows the native Go test format. Setting up the test environment is outside the scope of the framework, and is left to the user. The framework will however, detect the environment and run the tests accordingly.
```
Comment (nirs) [inline docs/e2e-test.md:7]: Very hard to read when lines are not wrapped. Please use short lines (72 chars works good for text).

File: docs/e2e-test.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+* Improved Reproducibility: A Go.work file allows us to explicitly define the versions of dependencies required for our tests. This ensures that all team members work with the same dependency versions, leading to consistent test results and reducing the likelihood of hard-to-debug issues stemming from dependency mismatches.
+* Simplified Dependency Updates: When updating dependencies, a Go.work file enables us to quickly identify and update test-specific dependencies without affecting the main project. This reduces the chances of inadvertently introducing breaking changes or compatibility issues.
+* Easier Onboarding: For new team members or external contributors, a Go.work file provides a clear separation of test dependencies from the main project. This simplifies the onboarding process and makes it easier for newcomers to understand the project structure and contribute effectively.
+* Enhanced Security: By isolating test dependencies, we minimize the potential attack surface in our main project. This approach helps ensure that vulnerabilities in test dependencies do not compromise the security of our production code.
```
Comment (nirs) [inline docs/e2e-test.md:19]: Isolating the e2e framework from ramen is nice - but we do really need go workspace? Can just have a separate go project in the e2e directory? Go will treat such project as part of ramen project?

Regardless, this text is really not useful for developers, so we can move it to the end of the doc.

The most important thing to be at the top is how to install this (hopefully nothing needed) and how to run the tests. Everything else can be at the end of the doc or in another doc.

File: docs/e2e-test.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+You can read more about go.work files here. https://go.dev/ref/mod#workspaces
+
+### Configuration
+
+
+The E2E test framework is organized into the following directories and files:
+
+* e2e/: The main directory containing the E2E test framework.
+* e2e/config/: The directory containing the configuration files for the E2E tests.
+* e2e/config/config.yaml: The main configuration file for the E2E tests.
```
Comment (nirs) [inline docs/e2e-test.md:30]: This is just the default configuration, not the main configuraztion.

File: docs/e2e-test.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+* e2e/go.mod: The Go module file containing the E2E test dependencies.
+* e2e/go.sum: The Go checksum file containing the E2E test dependencies' checksums.
+* e2e/main_test.go: The main test file where the test suites are defined.
```
Comment (nirs) [inline docs/e2e-test.md:33]: This is the most important info in this section, the rest is not really needed.

File: docs/e2e-test.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+The E2E test framework utilizes [Viper](https://github.com/spf13/viper) for loading and validating the configuration, and [Testify](https://github.com/stretchr/testify) for running test suites and performing validations.
+
+## Running the Tests
+
+## Running the Entire E2E Test
+To run the entire E2E test suite, use the following make target:
+
+```sh
+make test-e2e
```
Comment (nirs) [inline docs/e2e-test.md:43]: This is not enough - we need to document

- How to run specific tests (e.g. suites or some tests)
- How to run a test and stop on failure - no cleanup
- How to run a part of a complicated test
- How to add new suite
- How to add new test

Basically what we have in https://github.com/RamenDR/ramen/tree/main/test#testing-drenv - this framework should have the same functionality so we can delete basic-test and use only this framework.

File: Makefile (hunk 4 above)
Comment (nirs) [reply Makefile:176]: Can be nice to add something like:

```
test-e2e-binary:
    go test -o ramen-e2e .e2e
```

To make it easy to create a standalone ramen-e2e tool.

**Comment by ShyamsundarR** [comment]
@nirs I am assuming all comments need not be addressed to start the e2e framework and can be deferred as issues to track. Otherwise we will never get started!

File: e2e/config/config.yaml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+---
+Clusters:
+  hub:
+    kubeconfigpath: ./config/hubKconfig
+  c1:
+    kubeconfigpath: ./config/c1Kconfig
+  c2:
+    kubeconfigpath: ./config/c2Kconfig
```
[the remark this answers is not in the document]
Comment (nirs) [reply e2e/config/config.yaml:8]: #1015 creates config.yaml and kubeconfigs that can be used by e2e.

I did not find how to pass the config file path to go test, currently this works only by copying the config.yaml over the default config, or running go test from another directory, both are not good enough.

We something like:

    go test -run xxx -config /path/to/config

This is the basic requirement to allow developers to do the same thing we can do now with basic-test:

    test/basic-test/run test/envs/regional-dr.yaml

This is also required to run with external clusters - you got kubeconfigs from some clusters, you don't want overwrite the default config.yaml to point to the kubeconfigs - it is part of the source.

**Comment by nirs** [comment]
> @nirs I am assuming all comments need not be addressed to start the e2e framework and can be deferred as issues to track. Otherwise we will never get started!

If we defer them to issues they will never be done. If we want to use the e2e framework we need to address the issue snow.

We can merge what we have now to make some progress and allow contribution from others in this area, but we should not treat this the e2e framework as done before we have basic tests in place.

**Comment by ShyamsundarR** [comment]
> > @nirs I am assuming all comments need not be addressed to start the e2e framework and can be deferred as issues to track. Otherwise we will never get started!
> 
> If we defer them to issues they will never be done. If we want to use the e2e framework we need to address the issue snow.
> 
> We can merge what we have now to make some progress and allow contribution from others in this area, but we should not treat this the e2e framework as done before we have basic tests in place.

Agreed, hence I used the word "start" the e2e framework in my comment. My priority of events would be to get the basic framework and start adding tests, and in parallel improve the "glue" layers (like improve how we deal with config or makefiles etc.). Getting to the point where we can add tests is the highest priority at present, even if executing them is a bit cumbersome to start with.

**Pull request closed without merging**